social_core/backends/azuread_b2c.py

```diff
@@ -0,0 +1,181 @@
+
+MIT License
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+
+See https://nicksnettravels.builttoroam.com/post/2017/01/24/Verifying-Azure-Active-Directory-JWT-Tokens.aspx
+    for verifying JWT tokens.
+"""
+
+
+from cryptography.hazmat.primitives import serialization
+from jwt import DecodeError, ExpiredSignature, decode as jwt_decode
+from jwt.utils import base64url_decode
+import json
+import six
+
+try:
+    from jwt.algorithms import RSAAlgorithm
+except ImportError:
+    raise Exception('This backend (AzureADB2COAuth2) is only supported on Python 2 and Python 3.4+. '
+                    'Python 3.3 is not supported because of issues in Cryptography package in Python3.3.')
+
+from ..exceptions import AuthTokenError
+from .azuread import AzureADOAuth2
+
+
+class AzureADB2COAuth2(AzureADOAuth2):
+    name = 'azuread-b2c-oauth2'
+
+    BASE_URL = 'https://login.microsoftonline.com/{tenant_id}'
+    AUTHORIZATION_URL = '{base_url}/oauth2/v2.0/authorize'
+    OPENID_CONFIGURATION_URL = '{base_url}/v2.0/.well-known/openid-configuration?p={policy}'
+    ACCESS_TOKEN_URL = '{base_url}/oauth2/v2.0/token?p={policy}'
+    JWKS_URL = '{base_url}/discovery/v2.0/keys?p={policy}'
+
+    DEFAULT_SCOPE = ['openid', 'email']
+
+    EXTRA_DATA = [
+        ('access_token', 'access_token'),
+        ('id_token', 'id_token'),
+        ('refresh_token', 'refresh_token'),
+        ('id_token_expires_in', 'expires'),
+        ('exp', 'expires_on'),
+        ('not_before', 'not_before'),
+        ('given_name', 'first_name'),
+        ('family_name', 'last_name'),
+        ('tfp', 'policy'),
+        ('token_type', 'token_type')
+    ]
+
+    @property
+    def tenant_id(self):
+        return self.setting('TENANT_ID', 'common')
+
+    @property
+    def policy(self):
+        policy = self.setting('POLICY')
+
+        if not policy or not policy.startswith('b2c_'):
+            raise Exception('SOCIAL_AUTH_AZUREAD_B2C_OAUTH2_POLICY is required and should start with `b2c_`')
```

Comment:
This should be an `AuthException` or a more fitting subclass.

Reply:
Done